libtransact/src/context/manager/thread/handler.rs

```diff
@@ -0,0 +1,263 @@
+/*
+ * Copyright 2019 Cargill Incorporated
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ * -----------------------------------------------------------------------------
+ */
+
+use std::sync::mpsc;
+use std::sync::mpsc::{RecvError, Sender};
+use std::thread;
+
+use crate::context::error::ContextManagerThreadError;
+use crate::context::manager::thread::{
+    ContextManagerThread, ContextOperationMessage, ContextOperationResponse, ContextOperationResult,
+};
+use crate::context::ContextId;
+use crate::protocol::receipt::{Event, TransactionReceipt};
+use crate::state::Read;
+
+pub struct ContextManagerThreadHandler {
+    /// Handle to the Context Manager thread
+    thread_handle: Option<thread::JoinHandle<()>>,
+
+    /// Used to send ContextOperationMessage to the Context Manager thread
+    thread_sender: Sender<ContextOperationMessage>,
+}
+
+impl ContextManagerThreadHandler {
+    pub fn new(database: Box<dyn Read<StateId = String, Key = String, Value = Vec<u8>>>) -> Self {
+        let (thread_sender, thread_receiver) = mpsc::channel();
+        let handle = ContextManagerThread::new(database, thread_receiver).start();
+
+        ContextManagerThreadHandler {
+            thread_handle: Some(handle),
+            thread_sender,
+        }
+    }
+
+    pub fn shutdown(&mut self) -> Result<(), ContextManagerThreadError> {
```

Comment:
This should also be on a separate struct that is solely responsible for sending a shutdown message, since there can only be one join_handler, owned by the creator, but their can be multiple instances of the interface for context manager around the sender.